python/cuml/tsa/arima.pyx

```diff
@@ -47,244 +48,266 @@ from libc.stdlib cimport malloc, free
+                             .format(s))
+        if P + D + Q == 0 and s > 0:
+            raise ValueError("ERROR: Period specified for non-seasonal ARIMA:"
+                             " {}".format(s))
+        if d + D > 2:
+            raise ValueError("ERROR: Invalid order. Required: d+D <= 2")
+        if s != 0 and (p >= s or q >= s):
+            raise ValueError("ERROR: Invalid order. Required: s > p, s > q")
+        if p + q + P + Q + self.intercept == 0:
+            raise ValueError("ERROR: Invalid order. At least one parameter"
+                             " among p, q, P, Q and fit_intercept must be"
+                             " non-zero")
+        if p > 4 or P > 4 or q > 4 or Q > 4:
+            raise ValueError("ERROR: Invalid order. Required: p,q,P,Q <= 4")
+
+        # Get device array. Float64 only for now.
+        self.d_y, _, self.n_obs, self.batch_size, self.dtype \
+            = input_to_dev_array(y, check_dtype=np.float64)
+
+        if self.n_obs < d + s * D + 1:
+            raise ValueError("ERROR: Number of observations too small for the"
+                             " given order")
 
-        # Convert the lists to numpy arrays if needed
-        if type(mu) is list:
-            mu = np.array(mu)
-        if type(ar_params) is list:
-            ar_params = np.transpose(ar_params)
-        if type(ma_params) is list:
-            ma_params = np.transpose(ma_params)
-        self.mu, _, _, _, _ = input_to_host_array(mu)
-        self.ar_params, _, _, _, _ = input_to_host_array(ar_params)
-        self.ma_params, _, _, _, _ = input_to_host_array(ma_params)
-
-        # get host and device pointers. Float64 only for now.
-        h_y, h_y_ptr, n_samples, n_series, dtype = \
-            input_to_host_array(y, check_dtype=np.float64)
-        d_y, d_y_ptr, _, _, _ = input_to_dev_array(y, check_dtype=np.float64)
-
-        self.h_y = h_y
-        self.d_y = d_y
-        self.num_samples = n_samples
-        self.num_batches = n_series
-        self.yp = None
         self.niter = None  # number of iterations used during fit
 
     def __str__(self):
-        return "Batched ARIMA Model {}, mu:{}, ar:{}, ma:{}".format(
-            self.order, self.mu,
-            self.ar_params, self.ma_params)
+        if self.seasonal_order[3]:
+            return "Batched ARIMA{}{}_{}".format(self.order,
+                                                 self.seasonal_order[:3],
+                                                 self.seasonal_order[3])
+        else:
+            return "Batched ARIMA{}".format(self.order)
 
-    def _ic(self, ic_type):
+    @nvtx_range_wrap
+    def _ic(self, ic_type: str):
         """Wrapper around C++ information_criterion
         """
         cdef cumlHandle* handle_ = <cumlHandle*><size_t>self.handle.getHandle()
-        (p, d, q) = self.order
-        # Convert host parameters to device parameters (will be removed later)
+
+        cdef ARIMAOrder order
```

Comment:
maybe better to construct `ARIMAOrder` during the ctor itself and then just refer it as `self.order` everywhere else?

Reply:
@teju85 In Cython it is illegal to `cdef` the class member of a non `cdef`-ed class, and it is impossible to `cdef` `ARIMA` because it inherits from `Base`. Do you have a workaround in mind?

There are multiple pieces of code that I find very redundant, e.g the one that creates an `ARIMAParams` structure, but I don't see a good way to factorize that (can't define a Python member function that returns a C++ structure, etc). And it's about to get worse when adding support for fp32 because `ARIMAParams<float>` and `ARIMAParams<double>` are different types and a Python member cannot be templated.

Reply:
Not sure I understand your point @Nyrio . What I meant was something like this:
```
# in the ctor, create an object of ARIMAOrder type, via first constructing a local object of
# that type and then assigning it to the class member
def __int__(self, ...):
    ...
    cdef ARIMAOrder orderObj
    orderObj.p = p
    ....
    self.arimaOrder = orderObj

def _ic(self, ic_type: str):
    # now, at the place of usage, access this object like any others
    print("arima order p=%d" % self.arimaOrder.p)
    ....
```

Reply:
this also means that you don't have to necessarily store `self.order`, `self.seasonal_order` and `self.intercept` member variables as well.

Reply:
I understand what you mean, but as I said it is illegal in Cython to declare an attribute with a C++ type in a class that is not defined with `cdef`. When I tried something like your code above I got the error message:
```
Error compiling Cython file:
------------------------------------------------------------
...
        cdef ARIMAOrder cpp_order
        cpp_order.p, cpp_order.d, cpp_order.q = order
        cpp_order.P, cpp_order.D, cpp_order.Q, cpp_order.s = seasonal_order
        cpp_order.k = self.intercept
        self.cpp_order = cpp_order
                        ^
------------------------------------------------------------
cuml/tsa/arima.pyx:235:25: Cannot convert 'ARIMAOrder' to Python object
```

Reply:
I'll see if I can work around that with a Python class and a method that converts it to a C++ `ARIMAOrder` object but it will still have to be created every time, it would just remove redundancy

Reply:
Here's a completely worked example showing something similar.
### cpp.pyx
```# distutils: language = c++

cdef extern from "cpp.hpp":
  ctypedef struct Sample2:
    int a
    int b


class Test:
  def __init__(self, a, b):
    cdef Sample2 s2
    s2.a = a
    s2.b = b
    self.sample2 = s2

  def printVal(self):
    cdef Sample2 tmp = self.sample2
    print("a=%d b=%d" % (tmp.a, tmp.b))
```

### cpp.hpp
```struct Sample2 {
  int a, b;
};  // struct Sample2
```

### setup.py
```
from distutils.core import Extension, setup
from Cython.Build import cythonize
ext = Extension(name="cpp", sources=["cpp.pyx"])
setup(ext_modules=cythonize(ext))
```

### test.py
```
import cpp
t = cpp.Test(1, 2)
t.printVal()
```

Usage is: `python setup.py build_ext --inplace && python test.py`

Reply:
Your example works because the structure doesn't have any class method. If I remove the class methods from my declaration I can use the C++ object directly

Reply:
I can do that if you want, it will remove some redundancy and the only redundant part will be the Python `complexity` method redundant with the C++ one

Reply:
I see. Then yes, it makes sense to keep the python's `complexity` method.

Reply:
It's kinda weird, I checked how Cython stored the C++ structure and it turns out it converts it into a dictionary, and can then recreate the C++ structure from the dictionary on-demand:
```
>>> model.order
{'p': 1, 'd': 1, 'q': 1, 'P': 0, 'D': 0, 'Q': 0, 's': 0, 'k': 0}
```

I'm not a huge fan of this way of handling it because you can't write `self.order.p`, you need to recreate an `ARIMAOrder` object or use it as a dictionary. It would make more sense to create a namedtuple or smth like that... But anyway, that will do.